File: python/cmc.py
import numpy as np
from typing import Tuple

class Game:

    def __init__(self, seed=None):
        self.state = np.zeros((4,4), dtype=np.uint64)
        self._rng = np.random.default_rng(seed)

    def _new_tile(self):
        empty = np.argwhere(self.state == 0)
        new = self._rng.choice(empty, axis=0)
        if self._rng.random() < 0.8:
            self.state[new[0], new[1]] = 2
        else:
            self.state[new[0], new[1]] = 4

        return new

    def reset(self):
        self.__init__()
        self._new_tile()
        self._new_tile()
        return self.state

    def _rotate(self, n):
        self.state = np.rot90(self.state, n)

    def _compress(self):
        #up
        moved = False

        for x in range(4):
            for y in range(1, 4):
                #if not 0
                if self.state[y, x]:
                    new_y = y
                    while new_y > 0 and self.state[new_y - 1, x] == 0:
                        new_y -= 1
                    if new_y != y:
                        self.state[new_y, x] = self.state[y, x]
                        self.state[y, x] = 0
                        moved = True
        return moved

    def _merge(self):
        reward = 0
        for x in range(4):
            for y in range(3):
                #if not 0
                if self.state[y, x]:
                    if self.state[y, x] == self.state[y + 1, x]:
                        self.state[y + 1, x] = 0
                        self.state[y, x] *= 2
                        reward += int(self.state[y, x])
        return reward

    def _move(self, direction):
        """
        directions:
        0 - up
        1 - right
        2 - down
        3- left
        """
        moved = False

        self._rotate(direction)

        moved = self._compress()

        reward = self._merge()
        moved = moved or reward
        
        #or operator is short-circuit
        moved = self._compress() or moved

        self._rotate((4 - direction) % 4)
        return moved, reward


    def step(self, direction) -> Tuple[np.ndarray, int]:
        # move + generate tiles and check gameover

        # frontend should check this property ad give option to restart
        # to restart, simply call this function again
        if self.gameover:
            return self._reset(), 0

        moved, reward = self._move(direction)
        if moved:
            self._new_tile()
        return self.state, reward

    #property for compatibility with existing frontent
    @property
    def gameover(self):

        if not self.state.all():
            return False

        for y in range(3):
            for x in range(4):
                if self.state[y, x] == self.state[y+1, x]:
                    return False
        for y in range(4):
            for x in range(3):
                if self.state[y, x] == self.state[y, x+1]:
                    return False
        return True

if __name__ == "__main__":
    game = Game()
    for i in range(2):
        game._new_tile()
        game._new_tile()
    print(game.state)
    game._compress()
    # grid.rotate(2)
    print(game.state)

File: java/CMC.java
import java.util.ArrayList;
import java.util.Random;

import helpers.MoveReturn;
import helpers.RotateLongMatrix;
import helpers.Direction;
import helpers.GameInterface;

public class CMC implements GameInterface {
    private long[][] grid;
    private long score;
    private Random rng = new Random();

    public CMC() {
        // initalized with default value 0
        grid = new long[4][4];
        score = 0;
    }

    public void reset() {
        grid = new long[4][4];
        new_tile();
        new_tile();
        score = 0;
    }

    public long[][] getGrid() {
        return grid;
    }

    public long getScore() {
        return score;
    }

    private void new_tile() {
        // find all empty tiles
        var empty = new ArrayList<int[]>();
        for (int y = 0; y < 4; y++) {
            for (int x = 0; x < 4; x++) {
                if (grid[x][y] == 0) {
                    int[] val = {x, y};
                    empty.add(val);
                }
            }
        }

        var new_tile = empty.get(rng.nextInt(empty.size()));

        long new_val = 2;
        if (rng.nextLong() > 0.8) new_val = 4;
        grid[new_tile[0]][new_tile[1]] = new_val;
    }

    private void rotate_90() {
        RotateLongMatrix.rotate(grid);
    }

    private void rotate(int n) {
        for (int i = 0; i<n; i++) {
            rotate_90();
        }
    }

    private boolean compress() {
        boolean moved = false;
        for (int x = 0; x<4; x++) {
            for (int y = 1; y<4; y++) {
                if (grid[x][y] > 0) {
                    int new_y = y;
                    while (new_y > 0 && grid[x][new_y-1] == 0) {
                        new_y--;
                    }
                    if (new_y != y) {
                        grid[x][new_y] = grid[x][y];
                        grid[x][y] = 0;
                        moved = true;
                    }
                }
            }
        }

        return moved;
    }

    private long merge() {
        long reward = 0;

        for (int x = 0; x<4; x++) {
            for (int y = 0; y<3; y++) {
                if (grid[x][y] > 0 && grid[x][y] == grid[x][y+1]) {
                    grid[x][y+1] = 0;
                    grid[x][y] *= 2;
                    reward += grid[x][y];
                }
            }
        }

        return reward;
    }

    private MoveReturn move(Direction direction) {

        int rotate_num;
        switch (direction) {
            case Up: rotate_num = 3; break;
            case Right: rotate_num = 2; break;
            case Down: rotate_num = 1; break;
            case Left:
            default:
                rotate_num = 0; break;
        }
        rotate(rotate_num);

        boolean moved = compress();
        long reward = merge();
        moved = (compress() || moved || reward > 0);

        rotate((4 - rotate_num) % 4);

        return new MoveReturn(moved, reward);
    }

    public long step(Direction direction) {

        if (gameover()) {
            reset();
            return 0;
        }

        MoveReturn move_output = move(direction);

        if (move_output.moved()) {
            new_tile();
        }

        score += move_output.reward();

        return move_output.reward();
    }

    public boolean gameover() {

        for (int i = 0; i < 4; i++) {
            // for each val in grid[i]
            for (long val: grid[i]) {
                if (val == 0) {
                    return false;
                }
            }
        }

        for (int y = 0; y < 3; y++) {
            for (int x = 0; x < 4; x++) {
                if (grid[x][y] == grid[x][y+1]) {
                     return false;
                }
            }
        }

        for (int y = 0; y<4; y++) {
            for (int x = 0; x < 3; x++) {
                if (grid[x][y] == grid[x+1][y]) {
                    return false;
                }
            }
        }

        return true;
    }

}
